### ISSUE TITLE
WFOpt use_nonlocalpp_deriv feature status

### ISSUE THREAD

**Issue by ye-luo** [QMCPACK/qmcpack#3789]
**Is your feature request related to a problem? Please describe.**
In WFOpt, both kinetic energy and NLPP depends on WF optimizable parameters.

1. TWF::evaluateDerivatives compute both \partial_\alpha log(psi) and  \partial_\alpha KE.
so this function is currently a pure virtual function, namely, Each WFC must implement it. If a given WFC doesn't carry any optimization parameter. it still needs to implement a empty function to make zero contribution.

2. If use_nonlocalpp_deriv=yes, and NLPP algorithm=non-batched. This is a very slow code path due to repeatedly accept/reject moves.
* NLPP calls TWF::evaluateDerivativesWF which computes \partial_\alpha log(psi).
* Each WFC being part of TWF needs to implement WFC::evaluateDerivativesWF. The current base class implementation is runtime error stopper. The is intended to stop users and figure out if an zero contribution implementation is appropriate.
* We have meaningful implementation in MultiSlaterDetTableMethod, J2OrbitalSoA, J2OMPTarget, J1OrbitalSoA, J1Spin although the implmenetaiton was done quick and dirty.
* DiracDeterminantBase implements empty function to make tests without orbital optimization run. This is wrong when orbital optimization is enabled.

3. If use_nonlocalpp_deriv=yes, and NLPP algorithm=batched. This is a much faster code path.
* NLPP calls TWF::evaluateDerivRatios which computes \partial_\alpha log(psi(Rq)) log(psi(R)).
* Each WFC being part of TWF needs to implement WFC::evaluateDerivativesWF. The current base class implementation is zero contribution. I'm inclinde to change the base class implementation to runtime error. Convenience vs safety, I'm favoring safety.
* We have  meaningful implementation in J2OrbitalSoA, J2OMPTarget, J1OrbitalSoA only.
* Both single and multi Slater determinant implementations rely on base class zero contribution implementation. This is wrong when orbital optimization is enabled.

I'm inclinded to phasing out 2 and further develop 3.

We can change use_nonlocalpp_deriv to yes by default, the only breaking feature is MSD because NLPP algorithm=batched is default and case 3 adds zero contribution.

**Comment by prckent** [comment]
Are we still missing implementation for J3 with use_nonlocalpp_deriv=yes and NLPP algorithm=batched?

**Comment by ye-luo** [comment]
> Are we still missing implementation for J3 with use_nonlocalpp_deriv=yes and NLPP algorithm=batched?

Yes. It is still zero contribution now. I started a month ago but never got chance to finish it. Tests have been created.
